GoogleForCreators/web-stories-wp
# Improve AMP plugin compatibility

## Summary

This PR improves compatibility with the AMP plugin:

* Override the template mode to be Standard (AMP-First) when requesting a singular `web-story` or the Validated URL screen is being viewed for a `web-story`.
* Inject `web-stories` into the list of supported post types when requesting a web story. When not requesting a web story, exclude the `web-stories` post type from the list of supportable post types. This prevents the post type from appearing among the list of post types which can be enabled/disabled on the AMP Settings screen (as of v2.0).
* ~Adds workaround for `standalone="standalone"` being flagged as a validation error. This was resulting in the `body` element being emptied out when serving an `amp-story`. The AMP plugin needs to be updated to allow the attribute value to replicate the attribute name for such boolean attributes. See https://github.com/ampproject/amp-wp/issues/5156.~
* Prevents script handle collision for `amp-story-player` AMP component by renaming the handle used for the AMP Story block to be `standalone-amp-story-player`. This is only enqueued if not on an AMP page since otherwise it will be added automatically. This fixes a 404 error I saw in the editor: ![image](https://user-images.githubusercontent.com/134745/89215659-e6ead780-d57d-11ea-9388-793d310eddcb.png)


## Relevant Technical Choices

The changes in this PR are in lieu of there being a more centralized way of conditionally switching the template mode given the request. When/if that happens, much of the code in this PR can be revised. The closest issue for that presently is https://github.com/ampproject/amp-wp/issues/2817.

The logic in `Story_Post_Type::get_request_post_type()` is admittedly hacky!

## User-facing changes

“Stories” will no longer appear in the list of post types that can be enabled for AMP. 

## Testing Instructions

0. Activate the latest AMP plugin in the `develop` branch (2.0-beta2)
1. Create a story.
2. Upon saving the story, you should be able to see the story appear in the Validated URLs screen.
3. No validation errors should be reported.
4. Viewing the story on the frontend should show that the AMP Optimizer is running.

----

Fixes #3257
Fixes #1276
